Spigot-API-Patches/0199-Prioritise-own-classes-where-possible.patch

```diff
@@ -0,0 +1,132 @@
+  *api-version: 1.13
+  *
++ *classloader-prioritize-self: true<!-- Paper -->
++ *
+  *commands:
+  *  flagrate:
+  *    description: Set yourself on fire.
+@@ -241,6 +247,7 @@ public final class PluginDescriptionFile {
+     private PermissionDefault defaultPerm = PermissionDefault.OP;
+     private Set<PluginAwareness> awareness = ImmutableSet.of();
+     private String apiVersion = null;
++    private boolean classLoaderPrioritizeSelf = false; // Paper - prioritise own classes
+ 
+     public PluginDescriptionFile(@NotNull final InputStream stream) throws InvalidDescriptionException {
+         loadMap(asMap(YAML.get().load(stream)));
+@@ -927,6 +934,19 @@ public final class PluginDescriptionFile {
+         return apiVersion;
+     }
+ 
++    // Paper start - prioritise own classes
++    /**
++     * Returns whether the plugin's classloader should prioritize the classes
++     * of the plugin it loads for.
++     *
++     * @return whether the plugin's classloader should prioritize its own
++     * classes
++     */
++    public boolean shouldClassLoaderPrioritizeOwnClasses() {
++        return classLoaderPrioritizeSelf;
++    }
++    // Paper end
++
+     /**
+      * @return unused
+      * @deprecated unused
+@@ -1092,6 +1112,16 @@ public final class PluginDescriptionFile {
+         if (map.get("prefix") != null) {
+             prefix = map.get("prefix").toString();
+         }
++
++        // Paper start - prioritise own classes
++        if (map.get("classloader-prioritize-self") != null) {
++            try {
++                classLoaderPrioritizeSelf = (Boolean) map.get("classloader-prioritize-self");
++            } catch (ClassCastException exception) {
++                throw new InvalidDescriptionException(exception, "classloader-prioritize-self has wrong type");
++            }
++        }
++        // Paper end
+     }
+ 
+     @NotNull
+@@ -1161,6 +1191,8 @@ public final class PluginDescriptionFile {
+             map.put("prefix", prefix);
+         }
+ 
++        map.put("classloader-prioritize-self", classLoaderPrioritizeSelf); // Paper - prioritise own classes
++
+         return map;
+     }
+ 
+diff --git a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+index 3a02dbe9..3a6446df 100644
+--- a/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
++++ b/src/main/java/org/bukkit/plugin/java/PluginClassLoader.java
+@@ -107,7 +107,10 @@ public final class PluginClassLoader extends URLClassLoader { // Spigot
+         Class<?> result = classes.get(name);
+ 
+         if (result == null) {
+-            if (checkGlobal) {
++            String path = name.replace('.', '/').concat(".class"); // Paper - moved up
++            JarEntry entry = jar.getJarEntry(path); // Paper - moved up
++
++            if (checkGlobal && (entry == null && description.shouldClassLoaderPrioritizeOwnClasses())) { // Paper - prioritise own classes if they exist
```

Comment:
Commenting here so I don't forget for when I can be assed to do this :p 
`entry == null || !should`